Section on backporting (#66)
openjdk/guide: src/index.md

After a failure is handled by excluding a test, the main JBS issue should be re-triaged and possibly given a new priority. This should be handled by the standard triage process. A test exclusion results in an outage in our testing. This outage should be taken into consideration when triaging, in addition to the impact of the bug itself.

## Backing Out a Change

> If a change causes a regression that can't be fixed within reasonable time the best way to handle the regression can be to back out the change. Backing out means that the inverse (anti-delta) of the change is pushed to effectively undo the change in the repository. There are two parts to this task, how to do the bookkeeping in JBS, and how to do the actual backout in git or mercurial.

Reviewer: ```suggestion
If a change causes a regression that can't be fixed within reasonable time, the best way to handle the regression can be to back out the change. Backing out means that the inverse (anti-delta) of the change is pushed to effectively undo the change in the repository. There are two parts to this task, how to do the bookkeeping in JBS, and how to do the actual backout in git or mercurial.
```

> The backout is a regular change and will have to go through the standard code review process, but is considered a [trivial](#trivial) change. The rationale is that a backout is usually urgent in nature and the change itself is automatically generated. In areas where two reviewers are normally required, only one additional Reviewer is required for a backout since the person who is performing the backout also will review the change.

Reviewer 2: Recommend that both instances of "reviewer*" in the final sentence be of the same case.  I would choose upper.

Author: The first "reviewers" does not refer to two Reviewers. For instance in Hotspot two reviewers are required but only one of them have to be a Reviewer.

Reviewer 2: Ah! Thanks for the clarification.  The existing case make sense.

Date:   Wed Jun 23 20:31:32 2021 +0200

    My change

diff --git a/README.md b/README.md
index 399e7cc311f..4961acb2126 100644
--- a/README.md
+++ b/README.md
@@ -1,4 +1,4 @@
-# Welcome to the JDK!
+# Welcome to my modified JDK!

 For build instructions please see the
 [online documentation](https://openjdk.java.net/groups/build/doc/building.html),

$ git revert aa371b4f02c2f809eb9cd3e52aa12b639bed1ef5
[master d454489052d] Revert "My change"
 1 file changed, 1 insertion(+), 1 deletion(-)

$ git show d454489052dc6ff69a21ad9c8f56b67fdeb435ee
commit d454489052dc6ff69a21ad9c8f56b67fdeb435ee (HEAD -> master)
Author: Jesper Wilhelmsson <jesper.wilhelmsson@oracle.com>
Date:   Wed Jun 23 20:32:08 2021 +0200

    Revert "My change"

    This reverts commit aa371b4f02c2f809eb9cd3e52aa12b639bed1ef5.

diff --git a/README.md b/README.md
index 4961acb2126..399e7cc311f 100644
--- a/README.md
+++ b/README.md
@@ -1,4 +1,4 @@
-# Welcome to my modified JDK!
+# Welcome to the JDK!

 For build instructions please see the
 [online documentation](https://openjdk.java.net/groups/build/doc/building.html),
~~~

### How to work with mercurial when a change is backed out

> In order to backout a change, the `hg backout` command is recommended, which essentially applies the anti-delta of the change. Make sure you perform the backout in the most upstream repostiory the change has escaped to.

Reviewer 3: typo: repostiory

Author: Thank you for reviewing! Fixed the typo.

    merged result is left uncommitted.
~~~

# Backporting

::: {.box}
[Quick Links]{.boxheader}

* [Push approval for JDK updates](https://openjdk.java.net/projects/jdk-updates/approval.html)
* [Skara documentation on backports](https://wiki.openjdk.java.net/display/SKARA/Backports)
:::

Development of the latest version of the JDK often results in bug fixes that might be interesting to include in some of the JDK update releases still being maintained. Moving a fix from a more recent release train (e.g. JDK 17) to an older release train (e.g. JDK 11) is called *backporting*.

The guideline for what to backport into a specific update release will vary over the lifetime of that release. Initially more fixes are expected to be backported as new features and large changes introduced in a mainline release stabilize. Over time the focus will shift from stabilization to keeping it stable - the release will go into maintenance mode. This means that bug fixes that require larger disruptive changes are more likely to be made in mainline and backported to more recent release trains only, and not to older release trains.

> Over time it's likely that the code base will diverge between mainline and any given update release, and the cost of backporting will increase. The cost in this case is not only the effort needed to perform the actual backport, but also the cost inferred by bugs introduced by the backport. This should be taken into consideration when deciding if a change should be backported or not. For more details on how to reson around what to backport, [this email from JDK 8 Updates Project lead Andrew Haley](https://mail.openjdk.java.net/pipermail/jdk8u-dev/2020-June/012002.html) has some guidelines for JDK 8u. The reasoning in this mail is specific to JDK 8u, but will in general apply to any JDK release in maintenance mode.

Author: Nice catch! Fixed.

Any change that originally required a CSR will require a new CSR to be backported unless the backport was covered by the initial CSR. Changes to Java SE specifications cannot be made in an update release without a Java SE Maintenance Release. CSR-related issues affect interfaces and behavior and must be very carefully scrutinized.

## Working with backports in JBS

::: {.box}
[Terminology]{.boxheader}

Main issue - The top issue in a backport hierarchy. Eg. [JDK-8272373](https://bugs.openjdk.java.net/browse/JDK-8272373) is a main issue, while [JDK-8277498](https://bugs.openjdk.java.net/browse/JDK-8277498) and [JDK-8277499](https://bugs.openjdk.java.net/browse/JDK-8277499) are backport issues of this main issue.
:::

In general there is no need to create backport issues in JBS manually. All work that is done in JBS in preparation for a backport (requesting approvals etc) is done in the main issue. The backport issue will be created automatically by the bots when you integrate the change to the source code repository.

> There can be cases where it is motivated to create a backport issue before the fix is done. In these cases set the fix version of the backport to `N-pool`, where `N` is the release train the backport is targeting. E.g. `17-pool`

Reviewer 4: You might want to add a sentence saying that if a backport record is created ahead of time, all work done in JBS is still done in the main issue.

Author: Thank you for reviewing! Added the sentence.

> There can be cases where it is motivated to create a backport issue before the fix is done. In these cases set the fix version of the backport to `N-pool`, where `N` is the release train the backport is targeting. E.g. `17-pool`. Please note that even if a backport issue is created ahead of time, all work done in JBS is stil done in the main issue.

Reviewer 5: ```suggestion
There can be cases where it is motivated to create a backport issue before the fix is done. In these cases set the fix version of the backport to `N-pool`, where `N` is the release train the backport is targeting. E.g. `17-pool`. Please note that even if a backport issue is created ahead of time, all work done in JBS is still done in the main issue.
```

> There can be cases where it is motivated to create a backport issue before the fix is done. In these cases set the fix version of the backport to `N-pool`, where `N` is the release train the backport is targeting. E.g. `17-pool`. Please note that even if a backport issue is created ahead of time, all work done in JBS is still done in the main issue.

Author: Thank you for the review. I have included the change and CSR example so resolving this comment. Reply about n-pool in the next comment :-)

> Obviously it's possible to set the fix version to the exact release the backport is targeting, but this isn't recommended. When a change is pushed, the bots will look at the main issue as indicated in the PR title, and look for backports with the current `N.0.x` release version as fix version, if no such backport is found they will look for `N-pool`, and if that isn't found either, a new backport issue will be created. This means that if the backport has an exact fix version set, but is delayed and misses the release indicated by the fix version, a new backport issue is created with a small mess as the result.

Reviewer 6: Using the exact release e.g. `openjdk8u322` is the only option with 8u.

Author: I don't see why there should be a difference between 8u and later releases. The tooling handle these releases exactly the same way as far as I know. Is there any official communication saying that 8u shouldn't use 8-pool?

Setting the fix version of a backport to `N` is always wrong. JDK `N` has already been released (or you wouldn't be backporting to it) and can't get any more fixes.

## Requesting approvals for backports

> In order to be allowed to push a change to one of the OpenJDK Update Release repositories, an approval is required. The [official process for how to request push approval for a backport](https://openjdk.java.net/projects/jdk-updates/approval.html) describes in detail how to work with JBS when requesting approvals. In short, there's a label `jdk<release>u-fix-request` that should be added to the main JBS issue. Also put a motivation as to why the issue needs to be backported as a comment in the main issue. Once the label and motivation has been added, wait for the maintainers of the release to approve your request.

Reviewer 7: "wait for the maintainers of the release to approve your request" - Should this section describe that they do so by adding the label `jdk<release>u-fix-yes`?

Author: Yes. Thank you for reviewing! Added a sentence with the label.

> 

Reviewer 8: Hi,
I would propose to differ the development and release
repositories most update projects use (like jdk11u-dev and jdk11u).

In this case, in the first sentence, it should say 
"... OpenJDK Update **Development** repositories ..."

and at the end you could add
  If an update is in rampdown in the Release repository, 
  urgent fixes need to be approved with
  `jdk<release>u-critical-request` / `jdk<release>u-critical-yes`.

If you want to keep it short, omitting this is ok.  But this is 
also missing on the "JDK Updates: Requesting push approval for fixes"
wiki page referred to from here.

Reviewer 6: +1

Author: Thank you for these suggestions! I have included this.

Reviewer 8: +1 :)

## Using the Skara tooling to help with backports

> The Skara project developed support for backports. [The official Skara documentation](https://wiki.openjdk.java.net/display/SKARA/Backports) describes in detail how to work with the tooling to create backport PRs on GitHub or using the CLI tools. As described in the documentation, the [`/backport`](https://wiki.openjdk.java.net/display/SKARA/Commit+Commands#CommitCommands-/backport) command can be used to create the backport PR. If a backport PR is manually created, set the title to `Backport <original commit hash>`.

Reviewer 5: ```suggestion
The Skara tooling includes support for backports. [The official Skara documentation](https://wiki.openjdk.java.net/display/SKARA/Backports) describes in detail how to work with the tooling to create backport PRs on GitHub or using the CLI tools. As described in the documentation, the [`/backport`](https://wiki.openjdk.java.net/display/SKARA/Commit+Commands#CommitCommands-/backport) command can be used on a commit (not a PR!) to create the backport PR. If a backport PR is manually created, set the title to `Backport <original commit hash>`.
```

Reviewer 5: You did not like these suggestions? It has happened more than once that users try to use `/backport` on a PR, since commit commands (or even comments in general!) is not widely known about. I think at least that should be made more clear.

Author: These changes has been included in the current version. Thank you for your review and these suggestions!

Reviewer 5: I see that now. Thanks. It looked to me that they had not been included, but I must have just gotten lost in the Github UI.

> The Skara tooling includes support for backports. [The official Skara documentation](https://wiki.openjdk.java.net/display/SKARA/Backports) describes in detail how to work with the tooling to create backport PRs on GitHub or using the CLI tools. As described in the documentation, the [`/backport`](https://wiki.openjdk.java.net/display/SKARA/Commit+Commands#CommitCommands-/backport) command can be used on a commit (not a PR!) to create the backport PR. If a backport PR is manually created, set the title to `Backport <original commit hash>`.

Reviewer 9: Suggestion: "... set the PR title to `Backport <original commit hash>`. This ensures that the the bots will recognize it as a backport as opposed to a fix specifically targetting an older release only. One can tell whether or not the bots recognized a PR as a backport by the `backport` label being added if it is being recognized."

Reviewer 6: +1

Author: Thank you for this suggestion. Included.
